enhancements/ingress/route-subdomain.md

```diff
@@ -0,0 +1,642 @@
+example, suppose the cluster administrator has created the following
+IngressController:
+
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: IngressController
+metadata:
+  name: shard1
+  namespace: openshift-ingress-operator
+spec:
+  domain: shard1.apps.mycluster.com
+  routeSelector:
+    matchLabels:
+      shard: shard1
+```
+
+The application developer specifies a Route as follows:
+
+```yaml
+apiVersion: route.openshift.io/v1
+kind: Route
+metadata:
+  labels:
+    shard: shard1
+  name: hello-openshift
+  namespace: hello-openshift
+spec:
+  spec:
+    # host left unspecified.
+    subdomain: hello
+  # ...
+```
+
+Because the subdomain field is specified, the API will not set a default value
+for the host field.  Instead, the router deployment associated with the "shard1"
+IngressController will add an entry to the Route's status to indicate the host
+name under which the router exposes the Route:
+
+```yaml
+apiVersion: route.openshift.io/v1
+kind: Route
+metadata:
+  labels:
+    shard: shard1
+  name: hello-openshift
+  namespace: hello-openshift
+spec:
+  # ...
+  subdomain: hello
+  # ...
+status:
+  ingress:
+  - conditions:
+    - lastTransitionTime: # ...
+      status: "True"
+      type: Admitted
+    host: hello.shard1.apps.mycluster.com
+    routerCanonicalHostname: router-shard1.shard1.apps.mycluster.com
+    routerName: shard1
+    # ...
+```
+
+### API Extensions
+
+This enhancement uses the existing `spec.subdomain` field of the Route API.  It
+does not modify the structure or semantics of the API; it merely implements
+existing semantics that have been specified as optional and were previously
+unimplemented.
+
+### Risks and Mitigations
+
+The Route API's `spec.subdomain` field has existed since OpenShift 4.1 as an
+explicitly optional field; it has simply not been implemented before OpenShift
+4.11.  Users may already be using manifests that specify `spec.subdomain` on
+clusters running older OpenShift releases.  When these manifests are
+instantiated on a cluster where `spec.subdomain` is implemented, the resulting
+Route will behave differently.  To mitigate confusion, this enhancement will
+include a release note advising users of that change.
+
+Users can proactively check for manifests that specify `spec.subdomain` prior to
+upgrading to a version of OpenShift that supports `spec.subdomain`:
+
+```shell
+grep -e subdomain: -- manifests/*.yaml
+```
+
+Users and admins can also proactively check for instantiated Routes that specify
+`spec.subdomain` on a running cluster:
+
+```shell
+oc get routes --output=go-template --template='{{range .items}}{{if .spec.subdomain}}{{printf "%s %s %s\n" .metadata.namespace .metadata.name .spec.subdomain}}{{end}}{{end}}' --all-namespaces
+```
+
+However, per the Route API's existing semantics, a Route object that has already
+had `spec.host` defaulted on an older OpenShift cluster will continue to be
+served using the same host name, and `spec.subdomain` will be ignored; the
```

Comment:
I understand the API is a little confusing.  Unfortunately that arises from the necessity to maintain backwards compatibility.  Moreover the `spec.subdomain` field and these semantics have actually been in openshift/api for many releases—we just haven't implemented the documented semantics—so I'm reluctant to try to change the semantics now.  

It might be worth exploring the possibility of printing a warning when both `spec.host` and `spec.subdomain` are set.  I can look into what would be required to do that.  Maybe (although I'm more dubious of this possibility) we could even add an admission check that rejects _new_ routes (whilst allowing updates to old routes, for compatibility) that specify both `spec.host` and `spec.subdomain`.

Reply:
I added a warning in openshift-apiserver; see <https://github.com/openshift/openshift-apiserver/pull/254#issuecomment-1108076241>.